nautobot/docs/development/best-practices.md

```diff
@@ -63,3 +63,193 @@ class ExampleModel(PrimaryModel):
+    - In some cases simply excluding certain fields would be the next most preferable e.g. `Meta.exclude = ["unwanted_field", "other_unwanted_field"]`
+    - Finally, the last resort should be explicitly declaring the desired fields using `Meta.fields =`. This should be avoided because it incurs the highest technical debt in maintaining alignment between model fields and filters.
+- In the event that fields do need to be customized to extend lookup expressions, a [dictionary of field names mapped to a list of lookups](https://django-filter.readthedocs.io/en/stable/ref/filterset.html#declaring-filterable-fields) **may** be used, however, this pattern is only compatible with explicitly declaring all fields, which should also be avoided for the common case. For example:
+
+```python
+class UserFilter(NautobotFilterSet):
+    class Meta:
+        model = User
+        fields = {
+            'username': ['exact', 'contains'],
+            'last_login': ['exact', 'year__gt'],
+        }
+```
+
+- It is acceptable that default filter mappings **may** need to be overridden with custom filter declarations, but [`filter_overrides`](https://django-filter.readthedocs.io/en/stable/ref/filterset.html#customise-filter-generation-with-filter-overrides) (see below) should be used as a first resort.
+- Custom filter definitions **must not** shadow the name of an existing model field if it is also changing the type.
+    - For example `DeviceFilterSet.interfaces` is a `BooleanFilter` that is shadowing the `Device.interfaces` related manager. This introduces problems with automatic introspection of the filterset and this pattern MUST be avoided.
+- Custom filter definitions **may** shadow the name of an existing field IF this is to adapt a filter for a human-readable field value (such as `slug`) on a related field vs. the `pk` (the default for most related fields). This pattern still relies on accessing the related field by name and then traversing into the relationship using a nested lookup.
+
+### Filter Naming and Definition
+
+- Boolean filters for membership **must** be named with `has_{related_name}` (e.g. `has_interfaces`)
+
+- Boolean filters for identity **must** be named with `is_{name}` (e.g. `is_virtual_chassis`) although this is semantically identical to `has_` filters, there may be occasions where naming the filter `is_` would be more intuitive.
+
+- Filters **must** declare [`field_name`](https://django-filter.readthedocs.io/en/stable/ref/filters.html#field-name) when they have a different name than the underlying field they are referencing. Where possible the suffix component of the filter name should map directly to the underlying field name.
+
+  For example, `DeviceFilterSet.has_console_ports` could be better named, to assert that the filter name following the `has_` prefix is a one-to-one mapping to the underlying model's related field name (`consoleports`) which also means that `field_name` is not required:
```

Comment:
It's not clear to me why `field_name` would not be required for `has_consoleports`, since it *is* a different name than the underlying `consoleports` field - can you elaborate?

Reply:
You're right. I done confused myself.